consensus/event_loop.go

```diff
@@ -0,0 +1,908 @@
+// when to run the super majority vote check, and to signal the server that we can construct a QC.
+//
+// Reference implementation:
+// https://github.com/deso-protocol/hotstuff_pseudocode/blob/6409b51c3a9a953b383e90619076887e9cebf38d/fast_hotstuff_bls.go#L756
+func (fc *fastHotStuffEventLoop) ProcessValidatorVote(vote VoteMessage) error {
+	// Grab the event loop's lock
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Ensure the event loop is running. This guarantees that the chain tip and validator list
+	// have already been set.
+	if fc.status != eventLoopStatusRunning {
+		return errors.New("FastHotStuffEventLoop.ProcessValidatorVote: Event loop is not running")
+	}
+
+	// Do a basic integrity check on the vote message
+	if !isProperlyFormedVote(vote) {
+		return errors.New("FastHotStuffEventLoop.ProcessValidatorVote: Malformed vote message")
+	}
+
+	// Check if the vote is stale
+	if isStaleView(fc.currentView, vote.GetView()) {
+		return errors.Errorf("FastHotStuffEventLoop.ProcessValidatorVote: Vote has a stale view %d", vote.GetView())
+	}
+
+	// Check if the public key has already voted for this view. The protocol does not allow
+	// a validator to vote for more than one block in a given view.
+	if fc.hasVotedForView(vote.GetPublicKey(), vote.GetView()) {
+		return errors.Errorf(
+			"FastHotStuffEventLoop.ProcessValidatorVote: validator %s has already voted for view %d",
+			vote.GetPublicKey().ToString(),
+			vote.GetView(),
+		)
+	}
+
+	// Check if the public key has already timed out for this view. The protocol does not allow
+	// for a validator to vote for a block in a view that it has already timed out for.
+	if fc.hasTimedOutForView(vote.GetPublicKey(), vote.GetView()) {
+		return errors.Errorf(
+			"FastHotStuffEventLoop.ProcessValidatorVote: validator %s has already timed out for view %d",
+			vote.GetPublicKey().ToString(),
+			vote.GetView(),
+		)
+	}
+
+	// Compute the value sha3-256(vote.View, vote.BlockHash)
+	voteSignaturePayload := GetVoteSignaturePayload(vote.GetView(), vote.GetBlockHash())
+
+	// Verify the vote signature
+	if !isValidSignatureSinglePublicKey(vote.GetPublicKey(), vote.GetSignature(), voteSignaturePayload[:]) {
+		return errors.New("FastHotStuffEventLoop.ProcessValidatorVote: Invalid signature")
+	}
+
+	// Cache the vote in case we need it for later
+	fc.storeVote(voteSignaturePayload, vote)
+
+	// Check if the crank timer has elapsed or the event loop has constructed a QC in the current view.
+	// If so, then there's nothing more to do.
+	if !fc.hasCrankTimerRunForCurrentView || fc.hasConstructedQCInCurrentView {
+		return nil
+	}
+
+	// Check if the vote is for the chain tip. If not, then there's nothing more to do.
+	if vote.GetBlockHash() != fc.tip.block.GetBlockHash() {
+		return nil
+	}
+
+	// Check if we have a super-majority vote for the chain tip.
+	if voteQCEvent := fc.tryConstructVoteQCInCurrentView(); voteQCEvent != nil {
```

Comment:
So I think if you have this here then the crank timer is kindof irrelevant right? Because if you're creating the QC as soon as the votes come in, there's no need to wait for the crank timer.

So here we have a choice:
1. We can either leave this here, in which case the network will run as fast as it possibly can in the steady-state, with votes being shaped into QCs as soon as enough of them have arrived.
2. We can remove this, in which case the crank timer is the one responsible for assembling votes into QCs.

I think in past discussions we have leaned toward (2) because it's more fair to nodes with slow connections to the current leader. But let me know if you'd like to revisit this.

Reply:
Nevermind I see you check the crank timer above.